src/evm-block-extractor/src/block_extractor.rs

```diff
@@ -0,0 +1,157 @@
+    ) -> Self {
+        Self {
+            rpc_url,
+            blockchain: Arc::new(Mutex::new(blockchain)),
+            rpc_batch_size,
+            request_time_out_secs,
+        }
+    }
+
+    pub async fn latest_block_number(&self) -> anyhow::Result<u64> {
+        let rpc_url = &self.rpc_url;
+        let client = EthJsonRcpClient::new(ReqwestClient::new(rpc_url.to_string()));
+        client.get_block_number().await
+    }
+
+    pub async fn collect_blocks(
+        &mut self,
+        blocks: impl Iterator<Item = u64>,
+        max_no_of_requests: usize,
+    ) -> anyhow::Result<()> {
+        let rpc_url = &self.rpc_url;
+        let mut tasks = Vec::new();
+        let delay = Duration::from_secs(1) / max_no_of_requests as u32;
+        let semaphore = Arc::new(Semaphore::new(max_no_of_requests));
+        let batch_size = self.rpc_batch_size as usize;
+
+        let client = EthJsonRcpClient::new(ReqwestClient::new(rpc_url.to_string()));
+
+        for block_number_u64 in blocks {
+            let request_time_out_secs = self.request_time_out_secs;
+            let client = client.clone();
+            let blockchain = Arc::clone(&self.blockchain);
+
+            let permit = Arc::clone(&semaphore).acquire_owned().await?;
+
+            let task: JoinHandle<anyhow::Result<()>> = tokio::spawn(async move {
+                let block_number = BlockNumber::Number(block_number_u64.into());
+
+                //throttle the requests
+                tokio::time::sleep(delay).await;
+
+                let result = tokio::time::timeout(
+                    Duration::from_secs(request_time_out_secs),
+                    client.get_full_block_by_number(block_number),
+                )
+                .await;
+
+                log::info!("block result: {:?}", result);
+
+                match result {
+                    Ok(Ok(block)) => {
+                        let mut blockchain = blockchain.lock().await;
+
+                        blockchain.insert_block(&block).await?;
+
+                        let transactions = block.transactions;
+                        for chunk in transactions.chunks(batch_size) {
+                            let tx_hashes = chunk.iter().map(|tx| tx.hash).collect::<Vec<_>>();
+
+                            let receipts =
+                                client.get_receipts_by_hash(tx_hashes, batch_size).await?;
+
+                            blockchain.insert_receipts(&receipts).await?;
+                        }
+                    }
+                    Ok(Err(e)) => {
+                        println!("Failed to get block {}: {:?}", block_number_u64, e);
+                    }
+                    Err(e) => {
+                        println!("Request for block {} timed out: {:?}", block_number_u64, e);
+                    }
+                }
+
+                drop(permit);
+                Ok(())
+            });
+
+            tasks.push(task);
+        }
+
+        for task in tasks {
```

Comment:
[minor] You could use `join_all`

Reply:
Yeah, we currently don't use the `futures` crate.

Reply:
If you don't use `join_all`, then the requests are performed sequentially instead of in parallel